File: clients/csharp/src/MustelaLog.Client.Core/Models/LogEventRecord.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace MustelaLog.Client.Core.Models;

/// <summary>
/// Repräsentiert ein Logevent aus der API.
/// Zeitfelder bleiben bewusst als Rohstrings erhalten.
/// </summary>
public sealed class LogEventRecord
{
    [JsonPropertyName("log_event_id")] public string? LogEventId { get; set; }
    [JsonPropertyName("tenant_id")] public string? TenantId { get; set; }
    [JsonPropertyName("tenant_key")] public string? TenantKey { get; set; }
    [JsonPropertyName("source_id")] public string? SourceId { get; set; }
    [JsonPropertyName("source_key")] public string? SourceKey { get; set; }
    [JsonPropertyName("source_name")] public string? SourceName { get; set; }
    [JsonPropertyName("environment_code")] public string? EnvironmentCode { get; set; }
    [JsonPropertyName("ingest_request_id")] public string? IngestRequestId { get; set; }
    [JsonPropertyName("source_event_id")] public string? SourceEventId { get; set; }
    [JsonPropertyName("occurred_at")] public string? OccurredAt { get; set; }
    [JsonPropertyName("observed_at")] public string? ObservedAt { get; set; }
    [JsonPropertyName("received_at")] public string? ReceivedAt { get; set; }
    [JsonPropertyName("ingested_at")] public string? IngestedAt { get; set; }
    [JsonPropertyName("severity_number")] public int SeverityNumber { get; set; }
    [JsonPropertyName("severity_text")] public string? SeverityText { get; set; }
    [JsonPropertyName("event_name")] public string? EventName { get; set; }
    [JsonPropertyName("event_category")] public string? EventCategory { get; set; }
    [JsonPropertyName("event_action")] public string? EventAction { get; set; }
    [JsonPropertyName("event_outcome")] public string? EventOutcome { get; set; }
    [JsonPropertyName("message_text")] public string? MessageText { get; set; }
    [JsonPropertyName("raw_payload_json")] public string? RawPayloadJson { get; set; }
    [JsonPropertyName("host_name")] public string? HostName { get; set; }
    [JsonPropertyName("service_name")] public string? ServiceName { get; set; }
    [JsonPropertyName("component_name")] public string? ComponentName { get; set; }
    [JsonPropertyName("module_name")] public string? ModuleName { get; set; }
    [JsonPropertyName("process_name")] public string? ProcessName { get; set; }
    [JsonPropertyName("process_pid")] public int? ProcessPid { get; set; }
    [JsonPropertyName("thread_id")] public string? ThreadId { get; set; }
    [JsonPropertyName("actor_user_id")] public string? ActorUserId { get; set; }
    [JsonPropertyName("actor_principal")] public string? ActorPrincipal { get; set; }
    [JsonPropertyName("session_hash_sha256")] public string? SessionHashSha256 { get; set; }
    [JsonPropertyName("client_ip")] public string? ClientIp { get; set; }
    [JsonPropertyName("server_ip")] public string? ServerIp { get; set; }
    [JsonPropertyName("trace_id")] public string? TraceId { get; set; }
    [JsonPropertyName("span_id")] public string? SpanId { get; set; }
    [JsonPropertyName("correlation_id")] public string? CorrelationId { get; set; }
    [JsonPropertyName("request_correlation_id")] public string? RequestCorrelationId { get; set; }
    [JsonPropertyName("attributes_json")] public string? AttributesJson { get; set; }
    [JsonPropertyName("classification_code")] public string? ClassificationCode { get; set; }
    [JsonPropertyName("retention_policy_code")] public string? RetentionPolicyCode { get; set; }
    [JsonPropertyName("retention_delete_after")] public string? RetentionDeleteAfter { get; set; }
    [JsonPropertyName("legal_hold_flag")] public bool LegalHoldFlag { get; set; }
    [JsonPropertyName("canonical_hash_sha256")] public string? CanonicalHashSha256 { get; set; }
    [JsonPropertyName("previous_hash_sha256")] public string? PreviousHashSha256 { get; set; }
    [JsonPropertyName("source_signature")] public string? SourceSignature { get; set; }
    [JsonPropertyName("signature_algorithm")] public string? SignatureAlgorithm { get; set; }
    [JsonPropertyName("tenant_scope_key")] public string? TenantScopeKey { get; set; }

    public string GetPrettyAttributesJson()
    {
        if (string.IsNullOrWhiteSpace(AttributesJson))
        {
            return string.Empty;
        }

        try
        {
            using var document = JsonDocument.Parse(AttributesJson);
            return JsonSerializer.Serialize(document.RootElement, new JsonSerializerOptions { WriteIndented = true });
        }
        catch
        {
            return AttributesJson ?? string.Empty;
        }
    }

    public string GetPrettyRawPayloadJson()
    {
        if (string.IsNullOrWhiteSpace(RawPayloadJson))
        {
            return string.Empty;
        }

        try
        {
            using var document = JsonDocument.Parse(RawPayloadJson);
            return JsonSerializer.Serialize(document.RootElement, new JsonSerializerOptions { WriteIndented = true });
        }
        catch
        {
            return RawPayloadJson ?? string.Empty;
        }
    }
}

File: clients/java/src/main/java/de/sasd/mustelalog/client/model/LogEventRecord.java
package de.sasd.mustelalog.client.model;

import java.util.LinkedHashMap;
import java.util.Map;

public final class LogEventRecord {
    private String logEventId = "";
    private String occurredAt = "";
    private String observedAt = "";
    private String receivedAt = "";
    private String ingestedAt = "";
    private Integer severityNumber;
    private String severityText = "";
    private String eventName = "";
    private String eventCategory = "";
    private String eventAction = "";
    private String eventOutcome = "";
    private String messageText = "";
    private Object attributesJson;
    private Object rawPayloadJson;
    private String sourceId = "";
    private String sourceKey = "";
    private String sourceName = "";
    private String sourceType = "";
    private String environmentCode = "";
    private String hostName = "";
    private String serviceName = "";
    private String componentName = "";
    private String moduleName = "";
    private String processName = "";
    private Integer processPid;
    private String threadId = "";
    private String actorUserId = "";
    private String actorPrincipal = "";
    private String sessionHashSha256 = "";
    private String clientIp = "";
    private String serverIp = "";
    private String traceId = "";
    private String correlationId = "";
    private String requestCorrelationId = "";
    private String classificationCode = "";
    private String retentionPolicyCode = "";
    private String canonicalHashSha256 = "";
    private String previousHashSha256 = "";
    private String sourceSignature = "";
    private String signatureAlgorithm = "";
    private Map<String, Object> rawFields = Map.of();

    public String getLogEventId() { return logEventId; }
    public void setLogEventId(String logEventId) { this.logEventId = safe(logEventId); }
    public String getOccurredAt() { return occurredAt; }
    public void setOccurredAt(String occurredAt) { this.occurredAt = safe(occurredAt); }
    public String getObservedAt() { return observedAt; }
    public void setObservedAt(String observedAt) { this.observedAt = safe(observedAt); }
    public String getReceivedAt() { return receivedAt; }
    public void setReceivedAt(String receivedAt) { this.receivedAt = safe(receivedAt); }
    public String getIngestedAt() { return ingestedAt; }
    public void setIngestedAt(String ingestedAt) { this.ingestedAt = safe(ingestedAt); }
    public Integer getSeverityNumber() { return severityNumber; }
    public void setSeverityNumber(Integer severityNumber) { this.severityNumber = severityNumber; }
    public String getSeverityText() { return severityText; }
    public void setSeverityText(String severityText) { this.severityText = safe(severityText); }
    public String getEventName() { return eventName; }
    public void setEventName(String eventName) { this.eventName = safe(eventName); }
    public String getEventCategory() { return eventCategory; }
    public void setEventCategory(String eventCategory) { this.eventCategory = safe(eventCategory); }
    public String getEventAction() { return eventAction; }
    public void setEventAction(String eventAction) { this.eventAction = safe(eventAction); }
    public String getEventOutcome() { return eventOutcome; }
    public void setEventOutcome(String eventOutcome) { this.eventOutcome = safe(eventOutcome); }
    public String getMessageText() { return messageText; }
    public void setMessageText(String messageText) { this.messageText = safe(messageText); }
    public Object getAttributesJson() { return attributesJson; }
    public void setAttributesJson(Object attributesJson) { this.attributesJson = attributesJson; }
    public Object getRawPayloadJson() { return rawPayloadJson; }
    public void setRawPayloadJson(Object rawPayloadJson) { this.rawPayloadJson = rawPayloadJson; }
    public String getSourceId() { return sourceId; }
    public void setSourceId(String sourceId) { this.sourceId = safe(sourceId); }
    public String getSourceKey() { return sourceKey; }
    public void setSourceKey(String sourceKey) { this.sourceKey = safe(sourceKey); }
    public String getSourceName() { return sourceName; }
    public void setSourceName(String sourceName) { this.sourceName = safe(sourceName); }
    public String getSourceType() { return sourceType; }
    public void setSourceType(String sourceType) { this.sourceType = safe(sourceType); }
    public String getEnvironmentCode() { return environmentCode; }
    public void setEnvironmentCode(String environmentCode) { this.environmentCode = safe(environmentCode); }
    public String getHostName() { return hostName; }
    public void setHostName(String hostName) { this.hostName = safe(hostName); }
    public String getServiceName() { return serviceName; }
    public void setServiceName(String serviceName) { this.serviceName = safe(serviceName); }
    public String getComponentName() { return componentName; }
    public void setComponentName(String componentName) { this.componentName = safe(componentName); }
    public String getModuleName() { return moduleName; }
    public void setModuleName(String moduleName) { this.moduleName = safe(moduleName); }
    public String getProcessName() { return processName; }
    public void setProcessName(String processName) { this.processName = safe(processName); }
    public Integer getProcessPid() { return processPid; }
    public void setProcessPid(Integer processPid) { this.processPid = processPid; }
    public String getThreadId() { return threadId; }
    public void setThreadId(String threadId) { this.threadId = safe(threadId); }
    public String getActorUserId() { return actorUserId; }
    public void setActorUserId(String actorUserId) { this.actorUserId = safe(actorUserId); }
    public String getActorPrincipal() { return actorPrincipal; }
    public void setActorPrincipal(String actorPrincipal) { this.actorPrincipal = safe(actorPrincipal); }
    public String getSessionHashSha256() { return sessionHashSha256; }
    public void setSessionHashSha256(String sessionHashSha256) { this.sessionHashSha256 = safe(sessionHashSha256); }
    public String getClientIp() { return clientIp; }
    public void setClientIp(String clientIp) { this.clientIp = safe(clientIp); }
    public String getServerIp() { return serverIp; }
    public void setServerIp(String serverIp) { this.serverIp = safe(serverIp); }
    public String getTraceId() { return traceId; }
    public void setTraceId(String traceId) { this.traceId = safe(traceId); }
    public String getCorrelationId() { return correlationId; }
    public void setCorrelationId(String correlationId) { this.correlationId = safe(correlationId); }
    public String getRequestCorrelationId() { return requestCorrelationId; }
    public void setRequestCorrelationId(String requestCorrelationId) { this.requestCorrelationId = safe(requestCorrelationId); }
    public String getClassificationCode() { return classificationCode; }
    public void setClassificationCode(String classificationCode) { this.classificationCode = safe(classificationCode); }
    public String getRetentionPolicyCode() { return retentionPolicyCode; }
    public void setRetentionPolicyCode(String retentionPolicyCode) { this.retentionPolicyCode = safe(retentionPolicyCode); }
    public String getCanonicalHashSha256() { return canonicalHashSha256; }
    public void setCanonicalHashSha256(String canonicalHashSha256) { this.canonicalHashSha256 = safe(canonicalHashSha256); }
    public String getPreviousHashSha256() { return previousHashSha256; }
    public void setPreviousHashSha256(String previousHashSha256) { this.previousHashSha256 = safe(previousHashSha256); }
    public String getSourceSignature() { return sourceSignature; }
    public void setSourceSignature(String sourceSignature) { this.sourceSignature = safe(sourceSignature); }
    public String getSignatureAlgorithm() { return signatureAlgorithm; }
    public void setSignatureAlgorithm(String signatureAlgorithm) { this.signatureAlgorithm = safe(signatureAlgorithm); }
    public Map<String, Object> getRawFields() { return rawFields; }
    public void setRawFields(Map<String, Object> rawFields) { this.rawFields = rawFields == null ? Map.of() : rawFields; }

    public Map<String, Object> toMap() {
        LinkedHashMap<String, Object> map = new LinkedHashMap<>();
        map.put("logEventId", logEventId);
        map.put("occurredAt", occurredAt);
        map.put("observedAt", observedAt);
        map.put("receivedAt", receivedAt);
        map.put("ingestedAt", ingestedAt);
        map.put("severityNumber", severityNumber);
        map.put("severityText", severityText);
        map.put("eventName", eventName);
        map.put("eventCategory", eventCategory);
        map.put("eventAction", eventAction);
        map.put("eventOutcome", eventOutcome);
        map.put("messageText", messageText);
        map.put("sourceId", sourceId);
        map.put("sourceKey", sourceKey);
        map.put("sourceName", sourceName);
        map.put("sourceType", sourceType);
        map.put("environmentCode", environmentCode);
        map.put("hostName", hostName);
        map.put("serviceName", serviceName);
        map.put("componentName", componentName);
        map.put("moduleName", moduleName);
        map.put("processName", processName);
        map.put("processPid", processPid);
        map.put("threadId", threadId);
        map.put("actorUserId", actorUserId);
        map.put("actorPrincipal", actorPrincipal);
        map.put("sessionHashSha256", sessionHashSha256);
        map.put("clientIp", clientIp);
        map.put("serverIp", serverIp);
        map.put("traceId", traceId);
        map.put("correlationId", correlationId);
        map.put("requestCorrelationId", requestCorrelationId);
        map.put("classificationCode", classificationCode);
        map.put("retentionPolicyCode", retentionPolicyCode);
        map.put("canonicalHashSha256", canonicalHashSha256);
        map.put("previousHashSha256", previousHashSha256);
        map.put("sourceSignature", sourceSignature);
        map.put("signatureAlgorithm", signatureAlgorithm);
        map.put("attributes", attributesJson);
        map.put("rawPayloadJson", rawPayloadJson);
        if (!rawFields.isEmpty()) map.put("rawFields", rawFields);
        return map;
    }

    private String safe(String value) { return value == null ? "" : value; }
}
